chainstate/src/detail/in_memory_block_tree.rs

```diff
@@ -0,0 +1,446 @@
+                    "Block {id} validity doesn't match {block_validity:?}",
+                    id = block_index.block_id(),
+                );
+            }
+        }
+    }
+}
+
+/// A "reference" to a single `BlockIndex` tree.
+pub struct InMemoryBlockTreeRef<'a> {
+    arena: &'a Arena<BlockIndex>,
+    root_id: NodeId,
+}
+
+impl<'a> InMemoryBlockTreeRef<'a> {
+    fn new(arena: &'a Arena<BlockIndex>, root_id: NodeId) -> Self {
+        Self { arena, root_id }
+    }
+
+    #[allow(unused)]
+    pub fn root_id(&self) -> NodeId {
+        self.root_id
+    }
+
+    pub fn node(&self, id: NodeId) -> Option<&'a Node<BlockIndex>> {
+        self.arena.get(id)
+    }
+
+    /// Return an iterator over the entire tree, for depth-first traversal.
+    /// A parent will always be visited before its children.
+    /// The first visited block will always be the root block.
+    pub fn iter_all_ids(&self) -> impl Iterator<Item = NodeId> + 'a {
+        self.root_id.descendants(self.arena)
+    }
+
+    /// Same as iter_all_ids, but the root is excluded.
+    pub fn iter_child_ids(&self) -> impl Iterator<Item = NodeId> + 'a {
+        self.iter_all_ids().skip(1)
+    }
+
+    pub fn iter_all_block_indices(&self) -> impl Iterator<Item = &'a BlockIndex> {
+        self.iter_all_ids().map(|node_id| {
+            self.arena
+                .get(node_id)
+                .expect("Node being iterated over must be present in the arena")
+                .get()
+        })
+    }
+
+    pub fn iter_child_block_indices(&self) -> impl Iterator<Item = &'a BlockIndex> {
+        self.iter_child_ids().map(|node_id| {
+            self.arena
+                .get(node_id)
+                .expect("Node being iterated over must be present in the arena")
+                .get()
+        })
+    }
+
+    pub fn root_block_index(&self) -> &'a BlockIndex {
+        self.arena
+            .get(self.root_id)
+            .expect("Inconsistent InMemoryBlockTree - root node is not in the arena")
+            .get()
+    }
+
+    /// Iterate over all nodes, depth first, and call the provided function on each node.
+    /// If the function returns false, the corresponding subtree will not be descended into.
+    pub fn for_all<E: std::error::Error>(
+        &self,
+        mut handler: impl FnMut(InMemoryBlockTreeRef<'a>) -> Result<bool, E>,
+    ) -> Result<(), E> {
+        indextree_utils::for_all_depth_first(self.arena, self.root_id, |node_id| {
+            handler(InMemoryBlockTreeRef::new(self.arena, node_id))
+        })
+    }
+}
+
+/// A single `BlockIndex` tree.
+pub struct InMemoryBlockTree {
+    arena: Arena<BlockIndex>,
+    root_id: NodeId,
+}
+
+impl InMemoryBlockTree {
+    fn new(arena: Arena<BlockIndex>, root_id: NodeId) -> InMemoryBlockTree {
+        Self { arena, root_id }
+    }
+
+    #[allow(unused)]
+    pub fn root_id(&self) -> NodeId {
+        self.root_id
+    }
+
+    pub fn as_ref(&self) -> InMemoryBlockTreeRef<'_> {
```

Comment:
Why not use the AsRef trait?

Reply:
> Why not use the AsRef trait?

AsRef would require returning a literal reference to some type. Here I need to return `InMemoryBlockTreeRef` by value.